subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/LatestModuleConflictResolver.java

```diff
@@ -62,22 +61,43 @@ public void select(ConflictResolverDetails<T> details) {
         // Work backwards from highest version, return the first candidate with qualified version and release status, or candidate with unqualified version
         List<Version> sorted = new ArrayList<>(matches.keySet());
         sorted.sort(Collections.reverseOrder(versionComparator));
+        T bestComponent = null;
+        T unqalifiedComponent = null;
         for (Version version : sorted) {
             T component = matches.get(version);
             if (!version.isQualified()) {
-                details.select(component);
-                return;
+                if (bestComponent == null) {
+                    details.select(component);
+                    return;
+                } else if (unqalifiedComponent == null) {
+                    unqalifiedComponent = component;
+                }
             }
-            if (releaseBias) {
+            // Only care about the first qualified version that matches
+            if (bestComponent == null) {
                 ComponentResolveMetadata metaData = component.getMetadata();
                 if (metaData != null && "release".equals(metaData.getStatus())) {
                     details.select(component);
                     return;
                 }
+                // Special case Maven snapshots
+                if (metaData instanceof MavenModuleResolveMetadata) {
+                    if (((MavenModuleResolveMetadata) metaData).getSnapshotTimestamp() != null || metaData.getModuleVersionId().getVersion().endsWith("-SNAPSHOT")) {
```

Comment:
Maybe put this test (combining both `if`s) into a `isMavenSnapshot()` method. Would make this quite complex piece of code easier to read.